modules/tinymce/src/core/main/ts/undo/Operations.ts

```diff
@@ -157,6 +185,18 @@ export const transact = (undoManager: UndoManager, locks: Locks, callback: () =>
   return undoManager.add();
 };
 
+export const transactAsync = (undoManager: UndoManager, locks: Locks, callback: () => void | Promise<void>): UndoLevel | null | Promise<UndoLevel | null> => {
+  endTyping(undoManager, locks);
+  undoManager.beforeChange();
+  const result = ignoreAsync(locks, callback);
+
+  if (result) {
+    return result.then(() => undoManager.add());
```

Comment:
Sorry, this is not how `extra` works. The two callbacks are contained within a _single_ `transact` wrapper.	Your nicely refactored code will need to be combined a bit more.